File: src/geometry/vector2d.cpp
/*
 * vector2d.cpp
 *
 *  Created on: 01/03/2015
 *      Author: Carlos F. M. Faruolo <5carlosfelipe5@gmail.com>felipe
 */

#include "geometry/vector2d.hpp"

#include <math.h>

bool Vector2D::operator ==(const Vector2D& v) const
{
	return v.x == x and v.y == y;
}

bool Vector2D::operator !=(const Vector2D& v) const
{
	return v.x != x or v.y != y;
}

/** Returns true if this vector coordinates equals the ones from the given vector */
bool Vector2D::equals(const Vector2D& vector) const
{
	return (*this) == vector;
}

/** Creates and returns a copy of this vector via copy constructor */
Vector2D Vector2D::clone() const
{
	return Vector2D(x, y);
}

/** Creates a string with this vector coordinates (x, y) */
string Vector2D::toString() const
{
	return string() + "(" + x + ", " + y + ")";
}

/** Creates and returns an array with this Vectors coordinates, in correct order.  */
double* Vector2D::getCoordinates() const
{
	double* coord = new double[2];
	coord[0] = x;
	coord[1] = y;
	return coord;
}

double Vector2D::operator ~() const
{
	return sqrt( x*x + y*y );
}

/** Returns the length/magnitude of this vector. */
double Vector2D::magnitude() const
{
	return ~(*this);
}

/** Return true if x = 0 and y = 0; */
bool Vector2D::isZero() const
{
	return x == 0 and y == 0;
}

/** Creates a vector with length 1 and same direction as this vector. In other words, a new vector that is a normalized version of this vector. Note that <b>the original vector remains unchanged</b>. */
Vector2D Vector2D::unit() const
{
	return isZero()? Vector2D() : Vector2D(x/magnitude(), y/magnitude());
}

/** Divides this vector's coordinates by its length/magnitude, normalizing it.
The returned object is the vector instance itself after normalization. */
Vector2D& Vector2D::normalize()
{
	if(not isZero())
	{
		double length = magnitude();
		if(length > 0)
		{
			x /= length;
			y /= length;
		}
	}
	return *this;
}

/** Creates and returns the opposite of this vector. In other words, returns a vector with same coordinates as this, but with changed signal. Note that <b>the original vector remains unchanged</b>. */
Vector2D Vector2D::opposite() const
{
	return Vector2D(-x, -y);
}

/** Changes the signal of this vector coordinates, effectively reflecting it.
<br> The returned object is <b>the vector instance itself</b> after reflection. */
Vector2D& Vector2D::reflect()
{
	x = -x;
	y = -y;
	return *this;
}

/** Creates and returns a vector that represents the sum of this vector and the given vector. Note that <b>the original vector remains unchanged</b>.*/
Vector2D Vector2D::sum(const Vector2D& v) const
{
	return Vector2D(x + v.x, y + v.y);
}

/** Adds to this vector the given vector. In other words, it performs an addition to this vector coordinates.
<br> The returned object is <b>the vector instance itself</b> after summation. */
Vector2D& Vector2D::add(const Vector2D& v)
{
	x += v.x;
	y += v.y;
	return *this;
}

/** Creates a vector that represents the difference/displacement of this vector and the given vector, in this order. It's useful to remember that vector subtraction is <b>not commutative</b>: a-b != b-a.
<br> Note that <b>the original vector remains unchanged</b>. */
Vector2D Vector2D::difference(const Vector2D& v) const
{
	return  v.opposite().add(*this);
}

/** Subtracts from this vector the given vector. In other words, it performs an subtraction to this vector coordinates.
It's useful to remember that vector subtraction is <b>not commutative</b>: a-b != b-a.
<br> The returned object is the <b>the vector instance itself</b> after subtraction. */
Vector2D& Vector2D::subtract(const Vector2D& v)
{
	x -= v.x;
	y -= v.y;
	return *this;
}

/** Creates a vector that represents the scalar multiplication of this vector by the given factor. Note that <b>the original vector remains unchanged</b>.*/
Vector2D Vector2D::times(double factor) const
{
	return Vector2D(x * factor, y * factor);
}

/** Multiply this vectors coordinates by the given factor. The returned object is <b>the vector instance itself</b> after multiplication.*/
Vector2D& Vector2D::scale(double factor)
{
	x *= factor;
	y *= factor;
	return *this;
}

/** Compute the distance between this vector and the given vector. In other words, returns the length/magnitude of the displacement between this and the given vector. */
double Vector2D::distance(const Vector2D& v) const
{
	return difference(v).magnitude();
}

/** Compute the inner/dot product between this and the given vector. */
double Vector2D::innerProduct(const Vector2D& v) const
{
	return x*v.x + y*v.y;
}

File: java/src/model/geometry/Vector2D.java
package model.geometry;

public class Vector2D extends Vector
{
	private static final long serialVersionUID = 4465655714316563553L;

	/** The coordinates */
	public double x, y;
	
	/** Creates a Vector2D with the given coordinates */
	public Vector2D(double x, double y)
	{
		this.x = x;
		this.y = y;
	}
	
	/** Creates a Vector2D with the (0, 0) coordinates */	
	public Vector2D()
	{
		this.x = 0;
		this.y = 0;
	}
	
	/** Copy constructor */
	public Vector2D(Vector2D v)
	{
		this.x = v.x;
		this.y = v.y;
	}
	
	/** Returns true if the given object is a instance of Vector2D and its coordinates equals the ones from this */
	@Override
	public boolean equals(Object object)
	{
		return object instanceof Vector2D? equals((Vector2D) object) : false;
	}
	
	/** Returns true if the given vector coordinates equals the ones from this */
	public boolean equals(Vector2D vector)
	{
		return vector.x == x && vector.y == y;
	}
	
	@Override
	public String toString()
	{
		return "(" + x + ", " + y + ")";
	}
	
	@Override
	public long getNumberOfDimensions() 
	{
		return 2;
	}

	@Override
	public Vector[] getCanonicalBase() 
	{
		return new Vector[]{ X_VERSOR.clone(), Y_VERSOR.clone() };
	}

	@Override
	public double[] getCoordinates()
	{
		return new double[]{x, y};
	}
	
	@Override
	public Vector2D clone()
	{
		return new Vector2D(x, y);
	}
	
	@Override
	public double magnitude()
	{
		return Math.sqrt( x*x + y*y );
	}
		
	@Override
	public boolean isZero()
	{
		return x == 0 && y == 0;
	}	

	@Override
	public Vector2D unit()
	{
		return isZero()? ZERO : new Vector2D(x/magnitude(), y/magnitude());
	}
	
	@Override
	public Vector2D normalize()
	{
		if(! isZero())
		{
			final double length = magnitude();
			x /= length;
			y /= length;
		}
		return this;
	}
	
	/** {@inheritDoc} In this case, returns a vector with coordinates (-this.x, -this.y). */
	@Override
	public Vector2D opposite()
	{
		return new Vector2D(-this.x, -this.y);
	}
	
	@Override
	public Vector2D reflect()
	{
		x = -x;
		y = -y;
		return this;
	}

	/** Changes the signal of this vector's X coordinate, effectively reflecting it relative to the Y axis. 
		The returned object is the vector instance itself after reflection. */
	public Vector2D reflectX()
	{
		x = -x;
		return this;
	}

	/** Changes the signal of this vector's Y coordinate, effectively reflecting it relative to the X axis. 
	The returned object is the vector instance itself after reflection. */
	public Vector2D reflectY()
	{
		y = -y;
		return this;
	}
		
	@Override
	public Vector2D sum(Vector v)
	{
		Vector2D v2d = (Vector2D) v;
		return new Vector2D(this.x + v2d.x, this.y + v2d.y);
	}
	
	@Override
	public Vector2D add(Vector v)
	{
		Vector2D v2d = (Vector2D) v;
		this.x += v2d.x;
		this.y += v2d.y;
		return this;
	}

	@Override
	public Vector2D difference(Vector v)
	{
		return  ((Vector2D) v).opposite().add(this);
	}
	
	@Override
	public Vector2D subtract(Vector v)
	{
		Vector2D v2d = (Vector2D) v;
		this.x -= v2d.x;
		this.y -= v2d.y;
		return this;
	}
	
	@Override
	public Vector2D times(double factor){
		return new Vector2D(this.x * factor, this.y * factor);
	}
	
	@Override
	public Vector2D scale(double factor)
	{
		x *= factor;
		y *= factor;
		return this;
	}
	
	@Override
	public double distance(Vector vector)
	{
		return this.difference(vector).magnitude();
	}
	
	@Override
	public double innerProduct(Vector v)
	{
		Vector2D v2d = (Vector2D) v;
		return this.x*v2d.x + this.y*v2d.y;
	}
	
	public double innerProduct(Vector2D v)
	{
		return this.x*v.x + this.y*v.y;
	}
	
	@Override
	public Vector2D vectorProduct(Vector[] vectors) 
	{
		return new Vector2D(-this.x, this.y);
	}
	
	
	public Vector2D reflection()
	{
		//TODO
		return null;
	}
	
	public Vector2D projection(Vector2D v)
	{
		return v.unit().scale(this.innerProduct(v)/(v.innerProduct(v)));
	}
	
	public Vector2D rejection()
	{
		//TODO
		return null;
	}
	
	public Vector2D rotation()
	{
		//TODO
		return null;
	}
	
	public Vector2D rotate()
	{
		//TODO
		return null;
	}
	
	public Vector2D perpendicular()
	{
		//TODO
		return null;
	}
	
	/** Represents the null/zero vector. It has coordinates (0, 0). */
	public static final Vector2D ZERO  = new Vector2D(), NULL_VECTOR=ZERO;
	
	/** A vector codirectional with the X axis, with length 1. */
	public static final Vector2D X_VERSOR = new Vector2D(1, 0);

	/** A vector codirectional with the Y axis, with length 1. */
	public static final Vector2D Y_VERSOR = new Vector2D(0, 1);
}
